### PR TITLE
Adding dropdown to DatasourceControl and ability to change datasource

### LINKED ISSUE #6175: Allow users to pick a different datasource from the explore view
Currently there is no way to switch the datasource from the explore view. Users would have to go back to sqllab to build another datasource.

### PR DESCRIPTION
**Pull request by michellethomas**
Adding a ChangeDatasourceModal to change the datasource from explore. I also added a dropdown so now when you click on the datasource it will still take you to the edit modal, but instead of the icons I added a dropdown that lists "Edit datasource", "Explore in Sqllab", and "Change datasource".

The label still acts as a button that takes you to the edit modal. And pressing on the arrow icon will take you to the dropdown.

Screenshots:
image
Hovering over the datasource:
image
Dropdown:
image

Here's what it would look like if we added it as an icon along side what's there now:
image

Fixes: #6175 
@mistercrunch @graceguo-supercat @kristw @hughhhh

### CODE REVIEW

File: superset/assets/src/datasource/ChangeDatasourceModal.jsx [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                this.setSearchRef(ref);
+              }}
+              type="text"
+              bsSize="sm"
+              value={this.state.filter}
+              placeholder={t('Search / Filter')}
+              onChange={this.changeSearch}
+            />
+          </div>
+          {this.state.loading && <Loading />}
```
Comment (hughhhh) [inline superset/assets/src/datasource/ChangeDatasourceModal.jsx]: (nit) deconstruct values in before the return:
```
const { loading } = this.state;

```

File: superset/views/datasource.py [hunk 2]
```diff
@@ -55,6 +55,13 @@ def save(self):
         db.session.commit()
         return self.json_response(data)
 
+    @expose('/get/<datasource_type>/<datasource_id>/')
+    @has_access_api
+    def get(self, datasource_type, datasource_id):
```
Comment (hughhhh) [inline superset/views/datasource.py]: 2 things:
- I would add a small if condition to make sure we are returning some data. 
- Can we get a small unit test here testing a valid/bad`datasource_type` and `datasource_id`

File: superset/assets/src/datasource/ChangeDatasourceModal.jsx [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+          that.setState({ loading: false });
+          getClientErrorObject(response).then(({ error }) => {
+            this.props.addDangerToast(error.error || error.statusText || error);
+          });
+        });
+    }
+  }
+
+  setSearchRef(searchRef) {
+    this.searchRef = searchRef;
```
Comment (hughhhh) [inline superset/assets/src/datasource/ChangeDatasourceModal.jsx]: Any reason you didn't just do this in the constructor vs. doing a function/bind?

File: superset/assets/src/datasource/ChangeDatasourceModal.jsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    this.onEnterModal = this.onEnterModal.bind(this);
+    this.selectDatasource = this.selectDatasource.bind(this);
+    this.changeSearch = this.changeSearch.bind(this);
+  }
+
+  onEnterModal() {
+    if (this.searchRef) {
+      this.searchRef.focus();
+    }
+    const that = this;
```
Comment (hughhhh) [inline superset/assets/src/datasource/ChangeDatasourceModal.jsx]: Why are using `that` var to reference `this`?

Comment (michellethomas) [reply superset/assets/src/datasource/ChangeDatasourceModal.jsx]: I moved some of this over from an old feature (that was still using ajax calls), but I wasn't sure why this was here. I asked Grace about it and it's no longer necessary so I've removed it.

**Comment by elibrumbaugh** [comment]
Thanks for sharing this @michellethomas! Some quick feedback:

It looks like `data source` and `dropdown` are seperate buttons but they've been pushed together.

The [react bootstrap example](https://react-bootstrap.netlify.com/components/dropdowns/#split-button-dropdowns) we looked at uses the component `SplitButton`.

- Could we use that component tweaked to fit that space or do we need to find a creative solution on how to handle this with the existing component?
- One option would be to remove the rounded corners only on one side of the `data source` and `dropdown`? Might be pretty tricky/hacky though.
- Could you push them closer together maybe?

One visual affordance that would help us here is the hover state like in the react bootstrap split button example. The two elements being side by side are distinguished by their hover states. I strongly suggest we implement some form of indication you're hovering over one side or the other of the split button.

**Comment by michellethomas** [comment]
I just realized that with a long datasource name the menu gets cut off on the right side so I'll need to fix this next week.

**Comment by michellethomas** [comment]
Thanks for the feedback @elibrumbaugh, the SplitButton was difficult to customize for the size, appearance that we want (because this datasource label should stay consistent with other control labels). I can evaluate making the changes you suggested, there is no component that makes this super easy to do so any kind of change here would be a customization.

**Comment by elibrumbaugh** [comment]
Let me know if it would help to brainstorm a solution together!

**Comment by michellethomas** [comment]
@elibrumbaugh I was able to get back to this and make the design changes you suggested. Let me know what you think. For the hover color I used a darker version of the label color, let me know if you have other suggestions.

superset_datasource_control

**Review by graceguo-supercat — APPROVED** [review]
LGTM

File: superset/assets/src/datasource/ChangeDatasourceModal.jsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        onExit={this.setSearchRef}
+        bsSize="lg"
+      >
+        <Modal.Header closeButton>
+          <Modal.Title>{t('Select a datasource')}</Modal.Title>
+        </Modal.Header>
+        <Modal.Body>
+          <div>
+            <FormControl
+              id="formControlsText"
```
Comment (kristw) [inline superset/assets/src/datasource/ChangeDatasourceModal.jsx]: Do we need an `id`?

File: superset/assets/src/datasource/ChangeDatasourceModal.jsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+              bsSize="sm"
+              value={filter}
+              placeholder={t('Search / Filter')}
+              onChange={this.changeSearch}
+            />
+          </div>
+          {loading && <Loading />}
+          {datasources && (
+            <Table
+              columns={['name', 'type', 'schema', 'connection', 'creator']}
```
Comment (kristw) [inline superset/assets/src/datasource/ChangeDatasourceModal.jsx]: Can make this a constant so in the case that `Table` does shallow comparison, it won't detect columns array as new value every time.

File: superset/assets/src/datasource/ChangeDatasourceModal.jsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+              className="table table-condensed"
+              data={datasources}
+              itemsPerPage={20}
+              filterable={['rawName', 'type', 'connection', 'schema', 'creator']}
```
Comment (kristw) [inline superset/assets/src/datasource/ChangeDatasourceModal.jsx]: same with `columns`

File: superset/assets/src/explore/components/controls/DatasourceControl.jsx [hunk 8]
```diff
@@ -46,68 +49,93 @@ class DatasourceControl extends React.PureComponent {
...
-              </Tooltip>
+              <Tooltip id={'error-tooltip'}>{t('Click to edit the datasource')}</Tooltip>
             }
           >
-            <a
-              href={`/superset/sqllab?datasourceKey=${this.props.value}`}
-              target="_blank"
-              rel="noopener noreferrer"
+            <div className="btn-group">
+              <Label onClick={this.toggleEditDatasourceModal} style={{ cursor: 'pointer' }} className="label-btn-hover">
```
Comment (kristw) [inline superset/assets/src/explore/components/controls/DatasourceControl.jsx]: Perhaps can move `cursor: pointer` to `css` as well.

**Comment by kristw** [comment]
I have a few minor suggestions but overall LGTM.

**Comment by michellethomas** [comment]
thanks @kristw responded to your comments and fixed an issue with error message, apparently sometimes getClientErrorObject can return `message` not `error` so made a change to account for that.

**Review by kristw — APPROVED** [review]

**Comment by michellethomas** [comment]
Here's the dropdown rebased with the + metadata button
superset_datasource_controlv2

**Pull request merged**